aptos-move/framework/aptos-framework/sources/fungible_asset.move

```diff
@@ -42,383 +47,507 @@ module aptos_framework::fungible_asset {
+    public fun initialize_arbitrary_wallet<T: key>(
+        constructor_ref: &ConstructorRef,
+        metadata: Object<T>,
+    ): Object<FungibleAssetWallet> {
+        let wallet_obj = &object::generate_signer(constructor_ref);
+        let metadata = object::convert<T, FungibleAssetMetadata>(metadata);
+        move_to(wallet_obj, FungibleAssetWallet {
+            metadata,
+            balance: 0,
+            allow_ungated_transfer: true,
+        });
+        move_to(wallet_obj,
+            FungibleAssetWalletEvents {
+                deposit_events: object::new_event_handle<DepositEvent>(wallet_obj),
+                withdraw_events: object::new_event_handle<WithdrawEvent>(wallet_obj),
+                set_ungated_transfer_events: object::new_event_handle<SetUngatedTransferEvent>(wallet_obj),
+            }
+        );
+
+        object::object_from_constructor_ref<FungibleAssetWallet>(constructor_ref)
     }
 
+    /// Withdraw `amount` of fungible asset from `wallet` by the owner.
+    public fun withdraw<T: key>(
+        owner: &signer,
+        wallet: Object<T>,
+        amount: u64,
+    ): FungibleAsset acquires FungibleAssetWallet, FungibleAssetWalletEvents {
+        assert!(object::owns(wallet, signer::address_of(owner)), error::permission_denied(ENOT_WALLET_OWNER));
+        assert!(ungated_transfer_allowed(wallet), error::invalid_argument(EUNGATED_TRANSFER_IS_NOT_ALLOWED));
+        withdraw_internal(object::object_address(&wallet), amount)
+    }
 
-    /// Return `allowed_ungated_transfer`.
-    public fun ungated_transfer_allowed(wallet: &Object<FungibleAssetWallet>): bool acquires FungibleAssetWallet {
-        borrow_fungible_asset(wallet).allow_ungated_transfer
+    /// Deposit `amount` of fungible asset to `wallet`.
+    public fun deposit<T: key>(
+        wallet: Object<T>,
+        fa: FungibleAsset,
+    ) acquires FungibleAssetWallet, FungibleAssetWalletEvents {
+        assert!(ungated_transfer_allowed(wallet), error::invalid_argument(EUNGATED_TRANSFER_IS_NOT_ALLOWED));
+        deposit_internal(wallet, fa);
     }
 
-    /// Mint the `amount` of fungible asset.
+    /// Mint the specified `amount` of fungible asset.
     public fun mint(
         ref: &MintRef,
         amount: u64,
-    ): FungibleAsset acquires FungibleAssetMetadata {
-        assert!(amount != 0, error::invalid_argument(EAMOUNT_CANNOT_BE_ZERO));
-        let metadata = verify(&ref.metadata);
-        increase_supply(&ref.metadata, amount);
+    ): FungibleAsset acquires FungibleAssetMetadata, FungibleAssetMetadataEvents {
+        assert!(amount > 0, error::invalid_argument(EAMOUNT_CANNOT_BE_ZERO));
+        let metadata = ref.metadata;
+        increase_supply(&metadata, amount);
+
+        let events = borrow_global_mut<FungibleAssetMetadataEvents>(object::object_address(&metadata));
+        event::emit_event(&mut events.mint_events, MintEvent { amount });
```

Comment:
Wouldn't emitting an event on Mint be a problem for parallelisation? As far as I remember it is counter-based.

Reply:
I was thinking of making mint/burn events optional but removed them for now since mints/burns can be inferred from deposit/withdraw events

Reply:
That's a good callout.